pantab/__init__.py

```diff
@@ -14,3 +14,58 @@
     "frames_to_hyper",
     "test",
 ]
+
+# We link against HyperAPI in a fun way: In Python, we extract the function
+# pointers directly from the Python HyperAPI. We pass those function pointers
+# over to the C module which will then use those pointers to directly interact
+# with HyperAPI. Furthermore, we check the function signatures to guard
+# against API-breaking changes in HyperAPI.
+#
+# Directly using HyperAPI's C functions always was and still is discouraged and
+# unsupported by Tableu. In particular, Tableau will not be able to provide
+# official support for this hack.
+#
+# Because this is highly brittle, we try to make the error message as
+# actionable as possible and guide users in the right direction.
+
+api_incompatibility_msg = """
+pantab is incompatible with the version of Tableau Hyper API installed on your
+system. Please upgrade both `tableauhyperapi` and `pantab` to the latest version.
+If doing so does not fix this issue, please file an issue at
+https://github.com/innobi/pantab/issues mentioning the exact pantab and HyperAPI
+versions which triggered this error.
+"""
+
+
+def _check_signature(f, sig):
+    func_type = ffi.typeof(f)
```

Comment:
It is a bit confusing to have ffi imported below yet referenced here. Is this really just `cffi` or is there something special about what is in the Hyper API?

Reply:
> It is a bit confusing to have ffi imported below yet referenced here
Agree. I will move this down further.

> Is this really just cffi or is there something special about what is in the Hyper API?

No, this `ffi` object here is not the plain `cffi` module which you could also obtain through `import cffi`. This `ffi` object here is the "pre-compiled" ffi object, i.e. the thing returned by

```
from cffi import FFI
ffibuilder = FFI()
ffibuilder.set_source(<HyperAPI specific stuff>)
ffi = ffibuilder.compile()
```

This means the `ffi` variable we have here represents the HyperAPI-specific C-functions, containing all the meta-data about all the functions exposed by HyperAPI

Reply:
The `_check_signature` was replaced by a `_get_hapi_function` function and moved below the `import`